### ISSUE TITLE
Should we handle the launch event in the ServiceWorker or an Open Window.

### ISSUE THREAD

**Issue by fallaciousreasoning** [WICG/file-handling#3]
Currently we haven't decided as to where to handle the event, there are advantages and disadvantages to both approaches.

## ServiceWorker
- If there are multiple app instances/tabs open you can decide what one to use (or open a new one)
- Might be slow: we have to spin up the SW before handling the event
- Could potentially be swallowed silently (why isn't my file opening?)

## Current Window
- This works well for the single instance app
- If there are multiple windows, how do we decide which one to use?
- What if it should be handled in a different one? We might have to send a message to the ServiceWorker.

## Opinions
I'm partial to sending the event to the service worker, and letting it decide what to do, as it is the only place you can inspect all app windows (and so this seems a natural place to handle it), though I understand some people have privacy concerns.

**Comment by fallaciousreasoning** [comment]
@marcoscaceres @aliams do you have any thoughts on this?

**Comment by marcoscaceres** [comment]
@fallaciousreasoning thanks for pinging me. Unfortunately, I've not kept up with this work or service workers in a long while, so I don't feel qualified to give an opinion :( I'd maybe drag in some of the folks from the Service Workers WG... @jakearchibald can probably recommend some additional folks that might be good to ask.

**Comment by aliams** [comment]
I think we should try to better understand the use case from a web developer's perspective and how valuable and useful it would be to have the extra control over which instance receives a file launch in a multi-instance app scenario. Do we have examples of apps that would want to be able to do this today that might be able to speak to this?

I would say that we can achieve both options above depending on web developer needs here.

We can have an activation event on the current window or in the service worker. In the case of the current window, it would be worthwhile investigating if having the last instance open a file for multi-instance app scenarios would suffice.

In either case, I would suggest having different types of activations, such as:
- an app launched as a result of the app's icon being clicked
- an app launched as a result of a file being opened whose file type the app is registered to handle
- an app launched as a result of a URI being accessed whose scheme name the app is registered to handle

**Comment by fallaciousreasoning** [comment]
# Mitigating Concerns with ServiceWorker LaunchEvents
I think we can mitigate some of the concerns with the ServiceWorker launch event.

## Slowness
If the browser isn't open, the time to spin  up the ServiceWorker is likely to be fairly small compared to the amount of time we'll need to spin up the browser, so it probably won't have much effect on this case. The one we're really concerned about is when the browser/app are open when we fire off the event and nothing happens. It still seems like this time is probably not going to be that significant compared to the I/O required to actually open the file.

Another potential cause of slowness would be that developers decide to do some heavy pre-processing in the ServiceWorker, before starting a client, so users don't get immediate feedback that something is happening. We could reduce this by putting a small timeout on the LaunchEvent (say, 1 second),  which would encourage developers to pass on the FileHandle as quickly as possible.

# Silence
In order to ensure that users actually see something happen we could make it a requirement that client.focus() is called in a FileHandler, and, if it is not, the browser selects an open client, or opens a new client. In addition, a UserAgent could choose to black list apps that don't open a client, to discourage bad behavior. 

# On Control and Instances
Off the top of my head:

An (Image|Document) editor probably wants to open each new file in a new client (as they do on Windows and Mac), but there shouldn't be multiple editors for a single file (so the existing client for the file should be focused).

VS Code opens each file in a new tab of the last active editor. However, folders are opened in a new Window.

## Different Activation Events
This seems like it might be a more appropriate discussion for the [sw-launch](https://github.com/WICG/sw-launch/) repository but I kind of envisioned some kind of discriminated type union

```ts
interface IconLaunch {
  cause: 'icon';
}

interface FileLaunch {
  cause: 'file';
  files: WritablefileHandles[];
}

interface UrlLaunch {
  cause: 'url';
  request: FetchAPIRequest;
}

type LaunchEvent = IconLaunch | FileLaunch | UrlLaunch;
```
where `cause` lets you infer what properties will be available, but I haven't thought about it that much (and I don't know if we can do something like that explicitly in WebIDLs).

**Comment by fallaciousreasoning** [comment]
PR #6 adds a section with these mitigations, what are your thoughts?

**Comment by comp615** [comment]
Wanted to piggy back on Ali's comments here. I like where the direction is heading. One event that has multiple causes seems ideal. Here's a note from a few months ago I sent about some file sharing and share v2 activate events and SW vs window handling.

I don't see a strong benefit of going through the SW, other than the ability to choose the target window (or to choose one at all instead of a new one). It seems like more boilerplate to have to catch it as a post to a fake endpoint, then somehow redirect the image data into the open client; getting the client to navigate. Nevertheless, this customizability may be very important to some apps.

Perhaps then, a compromise is to use an activate event to carry information.
- activate event is triggered for a specific origin
- if that origin has an associated service worker, see if it has an activate handler and pass it the event if so. (I realize this is poorly named).
- if the service worker does not have a handler, or default is not prevented, select the best candidate window and trigger the activate event on it (this can remove the boilerplate of having to get it to a window)
- if the window does not have a handler, refresh to the default URL or post the data or whatever
-if no widow is open, open a new one and repeat the previous two events (note we get into a bad place with lazy loading here on Twitter, may need a helper tool which many would say is already the SW, but it's what we do with Windows)

These thoughts sound similar to the discussion above, where there's some concern around always opening a window. Happy to answer more questions around this from the Twitter side of things. We have a fairly built out handler based on Windows UWP format, and it's not too bad. But lots of edge cases to figure out.

**Comment by fallaciousreasoning** [comment]
Wow, I quite like this. So the event kind of bubbles up until it gets handled? 

This could make the `launch` event super dumb (a good thing, I think). Its only responsibility would be to decide which client window to focus (which is the whole reason we have to have it in the service worker). Actual handling of the event would be done in the client, which would alleviate some of our concerns about doing heavy lifting in the SW.

Do you have some reasons why you'd like a single unified event in the client (I'm ready to be convinced)? If we just kept using whatever mechanisms are already in place in the client window, this would be a much more progressive enhancement (whereas a unified launch wouldn't work in old browsers).

I'll have a bit more of a think about this and discuss it with some of my teammates :)

@raymeskhoury anything to add?

**Comment by raymeskhoury** [comment]
Hey @comp615, thanks for chiming in! 

You touched on 2 different points:
**1) Have a single launch event with different types** launch events were originally designed only to capture navigations. At face value this seems limiting as there are other types of "activation" that are not navigations. But the clever thing that mgiuca@ recognized when designing it is that all current ways of activating an app are implemented as navigations. This includes register protocol handler, web share target as well as the proposed [shortcuts API](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/master/Shortcuts/explainer.md). This means that launch event is able to give power to apps over where invocations from these old APIs open.

It's an open question whether new types of activation (e.g. file handlers) would also be implemented as navigations that can be captured as a launch event or whether they should just be a new special type of launch event. There is more discussion on this in the explainer link above.

**2) Whether launch events should be able to be handled in a window** It seems like your main argument here is that it would be simpler for developers if they didn't have to worry about handling the event in a service worker. I can understand this. The main downside I see is that it expands the API surface. The other question is whether it's powerful enough to give a good user experience or whether developers would always just use the SW anyway. 

It sounds like UWP activation events are handled on the window and that's what Twitter uses? Does the experience work well? One positive about this approach is that we could start with a first level of the API which _only_ works in windows, which alleviates many concerns around bad behaving apps. We could then expand the capability to service workers at a later date.

I think we should probably continue this discussion on the [sw-launch repo](https://github.com/WICG/sw-launch/blob/master/explainer.md) since it's really related to that API?

**Comment by fallaciousreasoning** [comment]
Closing in favor of WICG/sw-launch#8
